File: src/boj/boj_4485.java
package boj;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class boj_4485 {
    static final int[] dr = {0, 1, 0, -1};
    static final int[] dc = {1, 0, -1, 0};

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int n;
        int t = 1;
        while ((n = Integer.parseInt(br.readLine())) != 0) {
            int[][] board = new int[n][n];
            for (int r = 0; r < n; r++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                for (int c = 0; c < n; c++) {
                    board[r][c] = Integer.parseInt(st.nextToken());
                }
            }

            bw.append(String.format("Problem %d: %d\n", t++, dijkstra(n, board)));
        }
        bw.close();
        br.close();
    }

    static int dijkstra(int n, int[][] board) {
        PriorityQueue<Point> pq = new PriorityQueue<>();
        pq.add(new Point(0, 0, board[0][0]));
        boolean[][] visited = new boolean[n][n];
        visited[0][0] = true;

        while (!pq.isEmpty()) {
            Point current = pq.poll();
            if (current.r == n - 1 && current.c == n - 1) {
                return current.d;
            }

            for (int d = 0; d < dr.length; d++) {
                int nr = current.r + dr[d];
                int nc = current.c + dc[d];
                if (nr < 0 || n <= nr || nc < 0 || n <= nc || visited[nr][nc]) {
                    continue;
                }
                visited[nr][nc] = true;
                pq.add(new Point(nr, nc, current.d + board[nr][nc]));
            }
        }

        return -1;
    }

    static class Point implements Comparable<Point> {
        int r;
        int c;
        int d;

        public Point(int r, int c, int d) {
            this.r = r;
            this.c = c;
            this.d = d;
        }

        @Override
        public int compareTo(Point o) {
            return Integer.compare(d, o.d);
        }
    }
}

File: src/boj/boj_4485.kt
package boj_4485

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

var n = 0
lateinit var map: Array<IntArray>

fun main() {
    val bufferedReader = BufferedReader(InputStreamReader(System.`in`))
    val bufferedWriter = BufferedWriter(OutputStreamWriter(System.out))
    var num = 1
    while (init(bufferedReader)) {
        bufferedWriter.append("Problem ${num++}: ${dijkstra()}\n")
    }
    bufferedWriter.flush()
    bufferedWriter.close()
    bufferedReader.close()
}

fun init(bufferedReader: BufferedReader): Boolean {
    n = bufferedReader.readLine().toInt()
    if (n == 0) {
        return false
    }

    map = Array(n) {
        with(StringTokenizer(bufferedReader.readLine())) {
            IntArray(n) {
                nextToken().toInt()
            }
        }
    }
    return true
}

fun dijkstra(): Int {
    val dx = arrayOf(1, -1, 0, 0)
    val dy = arrayOf(0, 0, 1, -1)
    val priorityQueue = PriorityQueue<Candidate>() { c1, c2 ->
        c1.compare(c2)
    }.apply { add(Candidate(0, 0, map[0][0])) }
    val distance = Array(n) { IntArray(n) { Int.MAX_VALUE } }.apply { this[0][0] = map[0][0] }

    while (priorityQueue.isNotEmpty()) {
        val candidate = priorityQueue.poll();
        if (candidate.x == n - 1 && candidate.y == n - 1) {
            return distance[n - 1][n - 1]
        }

        for (d in 0 until 4) {
            val nx = candidate.x + dx[d]
            val ny = candidate.y + dy[d]
            if (nx in 0 until n && ny in 0 until n && map[ny][nx] + candidate.distance < distance[ny][nx]) {
                distance[ny][nx] = candidate.distance + map[ny][nx]
                priorityQueue.add(Candidate(nx, ny, distance[ny][nx]))
            }
        }
    }
    return -1
}

class Candidate(val x: Int, val y: Int, val distance: Int) {
    fun compare(other: Candidate) = distance.compareTo(other.distance)
}